### ISSUE TITLE
Should the plugin work between channels that are not announced?

### ISSUE THREAD

**Issue by 21M4TW** [daywalker90/sling#6]
Hi,

I am trying to rebalance between some of my channels that are not announced ("private") and status_str says "1:NoRoutes". In the log I see "channel not found in graph!" and "could not find a route. Sleeping...". Should this worked between this type of channels, or not?
Thanks!

**Comment by daywalker90** [comment]
I've actually found a bug preventing rebalances after the first in this scenario, fix will be in next version.

**Comment by 21M4TW** [comment]
2024-10-21T11:57:36.800Z DEBUG   plugin-sling: 839427x2270x0/1: Spawning job.
2024-10-21T11:57:36.800Z DEBUG   plugin-sling: 000000x0000x0: target: 00000 sats
2024-10-21T11:57:36.800Z DEBUG   plugin-sling: 000000x0000x0/1: Candidates: 11111x1111x0, 22222x2222x0
2024-10-21T11:57:36.800Z UNUSUAL plugin-sling: 000000x0000x0/1: channel not found in graph!
2024-10-21T11:57:36.800Z INFO    plugin-sling: 000000x0000x0/1: could not find a route. Sleeping...
2024-10-21T11:57:36.800Z DEBUG   plugin-sling: 000000x0000x0/1: Starting sleeper for 600s

**Comment by daywalker90** [comment]
Can you show me the exact command you used to create this job?

**Comment by 21M4TW** [comment]
> Can you show me the exact command you used to create this job?

lightning-cli --conf etc/lightningd.conf sling-job -k scid=00000x0000x0 direction=pull amount=1***** maxppm=250 target=1 candidates='["11111x1111x0", "22222x2222x0"]' maxhops=20

(Obviously using my own channel's short IDs. The amount is 25% of the maximum receivable for channel 00000x0000x0). All involved channels are not announced and I don't have any announced channel on my node. I am basically trying to pull all the liquidity for channel 00000x0000x0 on my side. Initially I tried with the default value for maxhops with the same result. Thanks!

**Comment by 21M4TW** [comment]
Just tried the same thing with the rebalance.py plugin. With that other plugin it is able to find a route (I guess it uses a different routing mechanism that does not rely on the presence of remote aliases) and it tries executing the rebalancing. However it fails with the error:
htlc 3 failed from 2th node with code 0x400a (WIRE_UNKNOWN_NEXT_PEER)

**Comment by daywalker90** [comment]
Every plugin must have the alias, either local or remote to send or receive. Since you have some remote aliases missing there will be certain rebalances not working. So this makes total sense.

**Comment by 21M4TW** [comment]
> Every plugin must have the alias for private channels, either local or remote to send or receive. Since you have some remote aliases missing there will be certain rebalances not working. So this makes total sense.

I am not sure since with rebalance.py, the rebalancing that fails is when the incoming scid is for the only peer that has a remote alias? And it did not prevent rebalance.py from finding a route (it shows me the exact route that will be attempted), it is the actual transaction that fails.

**Comment by daywalker90** [comment]
Yes you need the remote alias for receiving so this makes sense. It finds a route with the normal scid but then fails the transaction because its using the scid and not the alias (im guessing).

**Comment by 21M4TW** [comment]
> Yes you need the remote alias for receiving so this makes sense. It finds a route with the normal scid but then fails the transaction because its using the scid and not the alias (im guessing).

But I am saying is that it is the other way around, the transaction that failed with rebalance.py is when I tried a receiving node that had a remote alias. It succeeded when I tried to rebalance between two channels that did not have any remote alias.

**Comment by 21M4TW** [comment]
Hmm, I just gave it a try again and it seems it is still not working on my side? Somewhat different output though. Trying 

lightning-cli sling-job -k scid='******x***x0' direction=push amount=2***** maxppm=1000 candidates='["******x****x0"]' maxhops=20

2025-04-26T00:15:32.529Z DEBUG   plugin-sling: Getting all channels in gossip_store...
2025-04-26T00:15:32.529Z DEBUG   plugin-sling: gossip_reader: offset:41360824
2025-04-26T00:15:32.529Z DEBUG   plugin-sling: gossip_reader: header error at 41361676:failed to fill whole buffer (not an actual error if buffer could not be filled)
2025-04-26T00:15:32.529Z DEBUG   plugin-sling: gossip_reader: gossip_store read in: 0ms
2025-04-26T00:15:32.529Z DEBUG   plugin-sling: gossip_reader: found updates:0 announcements:17709 amounts:17709 deletes/dying:0
2025-04-26T00:15:32.529Z DEBUG   plugin-sling: gossip_reader: post_processing_time: 17ms
2025-04-26T00:15:32.530Z DEBUG   plugin-sling: Reading gossip store done after 17ms!
2025-04-26T00:15:32.530Z INFO    plugin-sling: 75252 public channels in sling graph after 20ms!
2025-04-26T00:15:32.530Z DEBUG   plugin-sling: Got 4 local channels after 20ms!
2025-04-26T00:15:32.530Z INFO    plugin-sling: 6 private channels in sling graph after 25ms!
2025-04-26T00:15:32.530Z INFO    plugin-sling: Refreshed graph in 25ms!
2025-04-26T00:15:32.954Z DEBUG   plugin-sling: Peerchannels refreshed in 1ms
2025-04-26T00:15:33.959Z DEBUG   plugin-sling: Peerchannels refreshed in 1ms
2025-04-26T00:15:34.065Z DEBUG   plugin-sling: Peerchannels refreshed in 1ms
2025-04-26T00:15:34.065Z DEBUG   plugin-sling: Read 0 pull jobs and 1 push jobs in 1ms
2025-04-26T00:15:34.065Z DEBUG   plugin-sling: ******x***x0/1: Spawning job.
2025-04-26T00:15:34.065Z DEBUG   plugin-sling: ******x***x0: target: 5*****sats
2025-04-26T00:15:34.066Z DEBUG   plugin-sling: ******x***x0/1: Candidates:
2025-04-26T00:15:34.066Z INFO    plugin-sling: ******x***x0/1: No candidates found. Adjust out_ppm or wait for liquidity. Sleeping...
2025-04-26T00:15:34.066Z INFO    plugin-sling: ******x***x0/1: could not find a route. Sleeping...
2025-04-26T00:15:34.066Z DEBUG   plugin-sling: ******x***x0x0/1: Starting sleeper for 600s

+-------+--------------+--------------------------------------------------------------------+------------+-----------+----------+------------------+------------------+
| alias | scid         | pubkey                                                             | status_str | rebamount | w_feeppm | last_route_taken | last_success_reb |
+-------+--------------+--------------------------------------------------------------------+------------+-----------+----------+------------------+------------------+
| ********** | ******x***x0 | [pubkey] | 1:NoRoutes | 0         | 0        | Never            | Never            |
+-------+--------------+--------------------------------------------------------------------+------------+-----------+----------+------------------+------------------+

There is a large channel between the peers of the two private channels I am trying to rebalance. Using CLN v25.02.2 and sling v3.0.3. The transfer correspond

**Comment by daywalker90** [comment]
Seems like all candidates were filtered out. This can be for various reasons: 

- Not connected
- too many pending htlcs
- Missing Gossip (fees or alias)
- Error in previous Iteration ( this temporary)

I'm busy right now, but i can Look further in to this in a few days.

**Comment by 21M4TW** [comment]
> Seems like all candidates were filtered out. This can be for various reasons:
> 
>     * Not connected
> 
>     * too many pending htlcs
> 
>     * Missing Gossip (fees or alias)
> 
>     * Error in previous Iteration ( this temporary)
> 
> 
> I'm busy right now, but i can Look further in to this in a few days.

Are you relying on listchannels for the local channels? It looks like listchannels does not list local unannounced channels, only listpeerchannels does.

**Comment by 21M4TW** [comment]
The listchannels command does not do what its documentation says. I opened an issue https://github.com/ElementsProject/lightning/issues/8248 .

**Comment by daywalker90** [comment]
No i'm not using listchannels. I'm using listpeerchannels which include private channels and then read the gossip file directly for public channels.

**Comment by 21M4TW** [comment]
> No i'm not using listchannels. I'm using listpeerchannels which include private channels and then read the gossip file directly for public channels.

Ok thanks, so this is not the issue then...

**Comment by 21M4TW** [comment]
`2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Spawning job.
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0: target: 5XXXXXsats
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: found custom candidate 862035x1110x0
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: XXXXXXxXXXXx0 is not normal: Missing remote alias for private channel
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Candidates:
2025-04-30T01:26:10.365Z INFO    plugin-sling: XXXXXXxXXXx0/1: No candidates found. Adjust out_ppm or wait for liquidity. Sleeping...
2025-04-30T01:26:10.365Z INFO    plugin-sling: XXXXXXxXXXx0/1: could not find a route. Sleeping...
`

**Comment by 21M4TW** [comment]
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Spawning job.
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0: target: 517185sats
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: found custom candidate XXXXXXxXXXXx0
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: XXXXXXxXXXXx0 is not normal: Missing remote alias for private channel
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Candidates:
2025-04-30T01:26:10.365Z INFO    plugin-sling: XXXXXXxXXXx0/1: No candidates found. Adjust out_ppm or wait for liquidity. Sleeping...
2025-04-30T01:26:10.365Z INFO    plugin-sling: XXXXXXxXXXx0/1: could not find a route. Sleeping...
2025-04-30T01:26:10.365Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Starting sleeper for 600s

**Comment by 21M4TW** [comment]
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Spawning job.
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0: target: XXXXXXsats
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: found custom candidate XXXXXXxXXXXx0
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: XXXXXXxXXXXx0 is not normal: Missing remote alias for private channel
2025-04-30T01:26:10.364Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Candidates:
2025-04-30T01:26:10.365Z INFO    plugin-sling: XXXXXXxXXXx0/1: No candidates found. Adjust out_ppm or wait for liquidity. Sleeping...
2025-04-30T01:26:10.365Z INFO    plugin-sling: XXXXXXxXXXx0/1: could not find a route. Sleeping...
2025-04-30T01:26:10.365Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Starting sleeper for 600s

**Comment by daywalker90** [comment]
> lightning-cli sling-job -k scid='XXXXXXxXXXx0' direction=push amount=2XXXXX maxppm=1000 candidates='["XXXXXXxXXXx0"]' maxhops=20

I assume you used the same settings like here? push and 1 candidate? If so there is nothing i can do, it comes back to https://github.com/ElementsProject/lightning/issues/8110 :

> XXXXXXxXXXXx0 is not normal: Missing remote alias for private channel

Your peer is probably running an old version and so you are unable to rebalance in that direction.

**Comment by 21M4TW** [comment]
> > lightning-cli sling-job -k scid='XXXXXXxXXXx0' direction=push amount=2XXXXX maxppm=1000 candidates='["XXXXXXxXXXx0"]' maxhops=20
> 
> I assume you used the same settings like here? push and 1 candidate? If so there is nothing i can do, it comes back to [ElementsProject/lightning#8110](https://github.com/ElementsProject/lightning/issues/8110) :
> 
> > XXXXXXxXXXXx0 is not normal: Missing remote alias for private channel
> 
> Your peer is probably running an old version and so you are unable to rebalance in that direction.

Yes I used the same settings. I can rebalance with the rebalance plugin as it does not rely on these remote alias entries, but initially it had other issues with unannounced channels (was relying on listchannels that does not work with local unannounced channels). I was able to fix them though: https://github.com/lightningd/plugins/compare/master...21M4TW:plugins:fix_unannounced_channels

**Comment by daywalker90** [comment]
Please try 3.0.5 i have removed the need for aliases. I thought it was required but if you say it's not... I can't test it myself since i don't have private channels without aliases and i don't know a way to fake it on regtest.

**Comment by 21M4TW** [comment]
> Please try 3.0.5 i have removed the need for aliases. I thought it was required but if you say it's not... I can't test it myself since i don't have private channels without aliases and i don't know a way to fake it on regtest.

Hmm, it still complains about route finding:
`
2025-05-03T02:10:02.439Z DEBUG   plugin-sling: Read 0 pull jobs and 1 push jobs in 1ms
2025-05-03T02:10:02.439Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Spawning job.
2025-05-03T02:10:02.439Z DEBUG   plugin-sling: XXXXXXxXXXx0: target: 5XXXXXsats
2025-05-03T02:10:02.439Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: build_candidatelist: found custom candidate XXXXXXxXXXXx0
2025-05-03T02:10:02.439Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Candidates: XXXXXXxXXXXx0
2025-05-03T02:10:02.523Z INFO    plugin-sling: XXXXXXxXXXx0/1: could not find a route. Sleeping...
2025-05-03T02:10:02.523Z DEBUG   plugin-sling: XXXXXXxXXXx0/1: Starting sleeper for 600s
`
Maybe I should look at the code and try figuring what causes the error on my node when I have the chance. I am not familiar with rust though.

**Comment by daywalker90** [comment]
Ah sorry i forgot to remove the alias check during route calculations. Please give 3.0.6 a try

**Comment by daywalker90** [comment]
It should be fixed in 3.0.6+, if not let me know

**Comment by 21M4TW** [comment]
> It should be fixed in 3.0.6+, if not let me know

Thanks! I just tested it and it seems to be working now!
